src/dbnode/storage/index.go

```diff
@@ -1316,6 +1316,11 @@ func (i *nsIndex) Query(
 	sp.LogFields(logFields...)
 	defer sp.Finish()
 
+	// Check if exceeding query limits at very beginning of query path.
+	if err := i.opts.IndexOptions().QueryStats().UpdateDocs(0); err != nil {
```

Comment:
nit: you can pull this check up a couple levels and bail even earlier by it placing here: https://github.com/m3db/m3/blob/master/src/dbnode/storage/database.go#L828.

Reply:
Good call. Done